gateway/index.js

```diff
@@ -24,11 +24,44 @@ async function loggingMiddleware(resolve, root, args, context, info) {
 (async () => {
   const { schema, executor } = await gateway.load();
 
+  const schemaHash = printSchema(schema);
   const server = new ApolloServer({
-    schema: applyMiddleware(schema, {
+    schema: applyMiddleware(
+      // We create a schema that creates proxy resolvers from the executor,
+      // so you can wrap resolvers with graphql-middleware
+      wrapSchema({
+        schema,
+        executor: executionRequest => {
+          const operation = getOperationAST(executionRequest.document, executionRequest.operationName);
+          const queryStr = print(executionRequest.document);
+          const operationName = operation?.name?.value || "";
+          return executor({
+            request: {
+              query: queryStr,
+              variables: executionRequest.variables || {},
+              operationName,
+              extensions: executionRequest.extensions,
+              http: {
+                method: executionRequest.context?.req?.method,
+                url: executionRequest.context?.req?.url,
+                headers: new Headers(executionRequest.context?.req?.headers),
+              },
+            },
+            schema,
+            schemaHash,
+            context: executionRequest.context || {},
+            queryHash: queryStr,
+            document: executionRequest.document,
+            source: queryStr,
+            operationName,
+            operation,
+          });
+        },
+        batch: true,
```

Comment:
WOW. That's indeed crazy solution. I tried it and it indeed works.

Few questions,
- Does `getOperationAST` and `print` methods on lines 35-36 are heavy operations?
- Can you elaborate about the meaning of the `batch: true` flag?
- Do I have to use empty context function? In our use case we have a custom context method that does more things

Reply:
The cost of them depends on the size of the document.
About "batch", this creates proxying resolvers for each root field. And this means executor will be called for each field. This flag allows us to collect those calls and batch them for the entire request to avoid multiple calls when you have multiple root fields.

Reply:
So it sounds like this is going to create huge performance impact on each call.

About the batch you mean this?

```graphql
query XXX {
     query1 {
         x
     }
    query 2 {
        y
    }
}
```

In this case `batch: true` will cause one run of the executor instead of two?

Reply:
>In this case batch: true will cause one run of the executor instead of two?

Exactly.

Reply:
Thanks for the info @ardatan!

Do you think there is some other way to workaround this while keep using `@apollo/gateway` without impacting the performance that much?

Reply:
Not using GraphQL Middleware is the best choice here. As I said before, it relies on resolvers so we need `wrapSchema` to mimic the behavior just for GraphQL Middleware.
I'd strongly recommend to use sth Envelop/Yoga hooks to apply any middleware logic.

Reply:
Unfortunately Apollo Server integration of Envelop is quite limited and might not work as expected;
You can see the discussion here; https://github.com/apollographql/apollo-server/discussions/5541
Also keep in mind that even if you use the Envelop plugin for GraphQL Middleware, it still won't work with Apollo Gateway because of the same reasons. GraphQL Middleware plugin still uses the same "resolvers" since it uses the same `applyMiddleware` thing under the hood.

Reply:
Understood. So currently, there is no way to use Envelop with Apollo Server? Only with Yoga?

Reply:
Unfortunately it doesn't seem possible to use Envelop fully with Apollo Server :(